src/rdd/rdd.rs

```diff
@@ -999,30 +1000,12 @@ pub trait Rdd: RddBase + 'static {
         Self: Sized,
         Self::Item: Data + Ord,
     {
-        if num == 0 {
-            Ok(vec![])
-        } else {
-            let first_k_func = Fn!(move |partition: Box<dyn Iterator<Item = Self::Item>>|
-                -> Box<dyn Iterator<Item = BoundedMinPriorityQueue<Self::Item>>>  {
-                    let mut queue = BoundedMinPriorityQueue::new(num);
-                    partition.for_each(|item: Self::Item| queue.append(item));
-                    Box::new(std::iter::once(queue))
-            });
-
-            let queue = self
-                .map_partitions(first_k_func)
-                .reduce(Fn!(
-                    move |mut queue1: BoundedMinPriorityQueue<Self::Item>,
-                          queue2: BoundedMinPriorityQueue<Self::Item>|
-                          -> BoundedMinPriorityQueue<Self::Item> {
-                        queue1.merge(queue2);
-                        queue1
-                    }
-                ))?
-                .unwrap() as BoundedMinPriorityQueue<Self::Item>;
-
-            Ok(queue.into_vec_sorted())
-        }
+        Ok(self
+            .map(Fn!(|x| Reverse(x)))
+            .take_ordered(num)?
+            .into_iter()
+            .map(|x| x.0)
+            .collect())
     }
 
```

Comment:
@rajasekarv @iduartgomez This is the refactoring approach Raja suggested. Looks like it does exactly what's expected, and is much aligned with the Scala Spark. Learned quite a lot new of Rust, thank you both. 
I am also trying Nacho's approach in another branch, seems a bit more work needs to be done. Will submit another PR for reviewing if I could get it eventually working.
The remaining work is to change `Ord` to `PartialOrd`, however the `BinaryHeap` seems to only accept `T: Ord`. Will keep you posted. Cheers.

Reply:
I would take this solution (maybe just try the other to see if there are any significant performance differences or for fun which really shouldn't as per the conversation Raja and I had before, since there is no shuffling invovling it should be a single pass iterator in the end and I was too quick to check out).

Yes BinaryHeap requires `Ord` so cannot change that :( Will leave a couple more comments but this is pretty close to it :)
```